src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -1337,159 +1336,341 @@ public void setMirror()
+                                      })
+                                      .map(request -> (IRequest<? extends R>) request)
+                                      .iterator());
+    }
+
+    public boolean hasCitizenCompletedRequests(@NotNull final CitizenData data)
+    {
+        return !getCompletedRequests(data).isEmpty();
+    }
+
+    public ImmutableList<IRequest> getCompletedRequests(@NotNull final CitizenData data)
+    {
+        if (!citizensByCompletedRequests.containsKey(data.getId()))
         {
-            return location;
+            return ImmutableList.of();
         }
 
-        /**
-         * Gets the ColonyView that this building belongs to.
-         *
-         * @return ColonyView, client side interpretations of Colony.
-         */
-        public ColonyView getColony()
+        return ImmutableList.copyOf(citizensByCompletedRequests.get(data.getId()).stream()
+                                      .map(getColony().getRequestManager()::getRequestForToken).filter(Objects::nonNull).iterator());
+    }
+
+    public <R> ImmutableList<IRequest<? extends R>> getCompletedRequestsOfType(@NotNull final CitizenData citizenData, final TypeToken<R> requestType)
+    {
+        return ImmutableList.copyOf(getCompletedRequests(citizenData).stream()
+                                      .filter(request -> {
+                                          Set<TypeToken> requestTypes = ReflectionUtils.getSuperClasses(request.getRequestType());
+                                          return requestTypes.contains(requestType);
+                                      })
+                                      .map(request -> (IRequest<? extends R>) request)
+                                      .iterator());
+    }
+
+    public <R> ImmutableList<IRequest<? extends R>> getCompletedRequestsOfTypeFiltered(
+                                                                                        @NotNull final CitizenData citizenData,
+                                                                                        final TypeToken<R> requestType,
+                                                                                        final Predicate<IRequest<? extends R>> filter)
+    {
+        return ImmutableList.copyOf(getCompletedRequests(citizenData).stream()
+                                      .filter(request -> {
+                                          Set<TypeToken> requestTypes = ReflectionUtils.getSuperClasses(request.getRequestType());
+                                          return requestTypes.contains(requestType);
+                                      })
+                                      .map(request -> (IRequest<? extends R>) request)
+                                      .filter(filter)
+                                      .iterator());
+    }
+
+    public void markRequestAsAccepted(@NotNull final CitizenData data, @NotNull final IToken token) throws IllegalArgumentException
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orionminecraft.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)